src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -400,6 +402,33 @@ public virtual async Task<ConnectionCompleteParams> Connect(ConnectParams connec
             return completeParams;
         }
 
+        private async Task<ConnectionCompleteParams> TryOpenConnectionWithRetry(ConnectionInfo connectionInfo, ConnectParams connectionParams)
+        {
+            int counter = 0;
+            ConnectionCompleteParams response = null;
+            while (counter <= MaxServerlessReconnectTries)
+            {
+                response = await TryOpenConnection(connectionInfo, connectionParams);
+                // If a serverless database is sleeping, it will return this error number and will need to be retried.
+                // See here for details: https://docs.microsoft.com/en-us/azure/azure-sql/database/serverless-tier-overview?view=azuresql#connectivity
+                if (response?.ErrorNumber == 40613)
+                {
+                    Logger.Write(TraceEventType.Information, "Attempt to retry serverless DB connection no. " + (counter + 1) + " for " + connectionInfo.OwnerUri);
+                    counter++;
+                }
+                else
+                {
+                    // Every other response, we can stop.
+                    break;
+                }
+            }
+            if (response != null && counter == MaxServerlessReconnectTries)

```

Comment:
If you do keep this - the check is wrong. Counter will be MaxServerlessReconnectTries + 1 when it exits the loop if it hit the max retries.